arch/risc-v/src/k230/k230_rptun.c

```diff
@@ -0,0 +1,501 @@
+ * Name: rp_is_autostart
+ ****************************************************************************/
+
+static bool rp_is_autostart(struct rptun_dev_s *dev)
+{
+  return true;
+}
+
+/****************************************************************************
+ * Name: rp_is_master
+ ****************************************************************************/
+
+static bool rp_is_master(struct rptun_dev_s *dev)
+{
+  struct k230_rptun_dev_s *priv = as_k230_rptun_dev(dev);
+  return priv->master;
+}
+
+/****************************************************************************
+ * Name: rp_start
+ ****************************************************************************/
+
+static int rp_start(struct rptun_dev_s *dev)
+{
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rp_stop
+ ****************************************************************************/
+
+static int rp_stop(struct rptun_dev_s *dev)
+{
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rp_notify
+ ****************************************************************************/
+
+static int rp_notify(struct rptun_dev_s *dev, uint32_t vqid)
+{
+#ifdef CONFIG_K230_RPTUN_IPI
+  k230_ipi_signal(RPTUN_IPI_NOTIFY_RX);
+#else
+  struct k230_rptun_dev_s *priv = as_k230_rptun_dev(dev);
+
+  if (priv->master)
+    {
+      priv->shmem->seq_rmt++;
+    }
+  else
+    {
+      priv->shmem->seq_mst++;
+    }
+
+#endif /* CONFIG_K230_RPTUN_IPI */
+  return 0;
+}
+
+/****************************************************************************
+ * Name: rp_reg_cb
+ ****************************************************************************/
+
+static int rp_set_callback(struct rptun_dev_s *dev, rptun_callback_t cb,
+                           void *arg)
+{
+  struct k230_rptun_dev_s *priv = as_k230_rptun_dev(dev);
+
+  priv->callback = cb;
+  priv->arg      = arg;
+
+  rpinfo("%p, arg:%p\n", cb, arg);
+
+  return 0;
+}
+
+#ifdef CONFIG_K230_RPTUN_MASTER
+/****************************************************************************
+ * Name: rp_reset
+ ****************************************************************************/
+
+static void rp_reset(struct rptun_dev_s *dev, int value)
+{
+  /* Only do hard reset for big core */
+
+  if (k230_hart_is_big())
+    {
+      rperr("Can't reset little core!\n");
+    }
+  else
+    {
+      k230_hart_big_stop();
+      k230_hart_big_boot(0);
+    }
+}
+
+/****************************************************************************
+ * Name: rp_panic
+ ****************************************************************************/
+
+static void rp_panic(struct rptun_dev_s *dev)
+{
+#ifdef CONFIG_K230_RPTUN_IPI
+  k230_ipi_signal(RPTUN_IPI_REMOTE_PANIC);
+#else
+  rperr("TODO: ask remote to panic.\n");
+#endif
+}
+#endif
+
+#ifndef CONFIG_K230_RPTUN_IPI
+
+static void k230_rptun_work(void *arg)
```

Comment:
why not handle irq directly?

Reply:
Simply because I haven't collected enough information about the chip's IPI hardware yet, so using lockless mode for communications. As per OpenAMP, lockless single-writer-single-reader should work though not efficient.